### PR TITLE
[Analog Output]Support N sample

### PR DESCRIPTION
**Pull request by yawhuu**
### What 
1. Support writing multiple points at one time (N Sample). The write function has two use cases now:
    1. `write()` which writea single point at one time
    2. NEW: `write(list_of_values, sample_rate)` which write multiple points at one time
2. Add two examples for the new feature.
    1. add write multiple points of data to single channel in [AO_singleChannel.py](examples/analog/AO_singleChannel.py)
    2. add write multiple points of data to multiple channels in [AO_multipleChannels.py](examples/analog/AO_multipleChannels.py)
3. Add document: [Generating FPGA Clocks](../docs/NI_ELVIS_III_Understanding_Gaps.md)

### How
Add writing N sample to Analog Output `write()`.

### Tests
 - Test_AnalogOutput_WriteSingleChannel
    - test_WriteTensPointWithOneThousandSampleRate_ReturnExpectedReadBack
    - test_WriteTenPointsThenWriteSinglePointInBankB_ReturnExpectedReadBack
    - test_WriteLargeNumberOfValuesWithOneThousandSampleRate_DoesnotShowAnyError
 - Test_AnalogOutput_WriteTwoChannels
    - test_WriteSixteenPointsWithOneThousandSampleRate_ReturnExpectedReadBack
 - Test_AnalogOutput_WriteAssertion
    - test_WriteWithInvalidNumberOfArguments_ShowAssertion
    - test_WriteOneSampleWithInvalidValue_ShowAssertion
    - test_WriteNSampleWithInvalidValues_ShowAssertion
    - test_WriteNSampleWithInvalidSampleRate_ShowAssertion
    - test_PassSampleRateThatIsGreaterThanMax_ShowAssertion
    - test_PassSampleRateThatIsLessThanMin_ShowAssertion

### CODE REVIEW

File: source/nielvisiii/academicIO.py [hunk 1]
```diff
@@ -233,6 +234,9 @@ def __write_to_cnfg_register(bank, cnfg):
             while not(cnfg == self.cnfg[bank].read() and self.ready[bank].read()):
                 pass
 
+            # wait 500 us to apply the AI registers once configuration is modified. 12*1000/40 M = 300 us < 500 us
+            time.sleep(0.5)
```
Comment (yawhuu) [inline source/nielvisiii/academicIO.py]: If we does not wait here, sometimes the readback value is wrong in the following stimulation:
write -> read -> write -> read ...

File: source/nielvisiii/academicIO.py [hunk 2]
```diff
@@ -410,15 +420,54 @@ def __init__(self, *configuration):
...
+            
+        self.go = self.session.registers['AO.SYS.GO']
+        self.stat = self.session.registers['AO.SYS.STAT']
+        self.lsb_weights = 4882813.0 / 1E+9
+        self.offset = 0.0 / 1E+9
+        self.signed = True
+
+    def write(self, *args):
+        """
+        Writes values to one or more analog output channels. Use the
```
Comment (gaoruhao) [inline source/nielvisiii/academicIO.py]: mention about "wait until done"

File: docs/NI_ELVIS_III_Generating_FPGA_Clocks.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+In the previous figure, you toggle the clock signal on every rising edge to generate the 20 MHz clock signal and toggle the clock signal on every other rising edge to generate the 10 MHz clock signal.
+
+The frequency to generate on the hardware must meet the following requirements:
+ - The frequency must be divisible by 25 ns because the clock cycle is 25 ns and you count the number of rising edges to make changes. For example, you cannot generate a frequency of 25 MHz. The first achievable frequency below 40 MHz is 20 MHz. 
+ - The frequency must be within the range of approximately 610.35 Hz to 40 MHz. The hardware uses a 16-bit counter that counts from 0 to 65,535. You can calculate the lowest achievable frequency by the equation 1/(25 ns * 65536) ≈ 610.35 Hz. 
+
+
+You can use clock divisors to generate even lower frequencies for different I/O types. You can divide the base frequency by even numbers and use the generated clock to increment the counter. For example, with a clock divisor of 2, the lowest achievable frequency is 1/(2 * 25 ns * 65536) ≈ 305.17 Hz.
+
+#### The following equation calculates frequencies that you can generate for pulse width modulation (PWM) I/O.
```
Comment (ling-wang) [inline docs/NI_ELVIS_III_Generating_FPGA_Clocks.md]:
```suggestion
#### The following = calculates the frequencies that you want to generate for pulse width modulation (PWM) I/O.
``` This is a test. If I left this comment successfully, you will find some words in the above sentence are highlighted.

File: examples/analog/AO_multipleChannels.py [hunk 4]
```diff
@@ -43,3 +54,25 @@
...
+                              'channel': channel0},
+                             {'bank': bank,     # define second channel: AO1
+                              'channel': channel1}) as AO_multiple_channels:
+    # write to the AO channels 20 times
+    for i in range(0, 20):
+        # define a list of floating-point numbers
+        input_value = [1.0, 2.5]
+        # specifies the sampling frequency, in hertz, of the output signal.
+        sample_rate = 1000
+        # write 1.0 and 2.5 to AO0 and AO1 on bank A
```
Comment (yawhuu) [inline examples/analog/AO_multipleChannels.py]: write 1.0 and 2.5 to both AO0 and AO1.

File: source/nielvisiii/academicIO.py [hunk 5]
```diff
@@ -238,6 +239,9 @@ def __write_to_cnfg_register(bank, cnfg):
             while not(cnfg == self.cnfg[bank].read() and self.ready[bank].read()):
                 pass
 
+            # wait 500 us to apply the AI registers once configuration is modified. 12*1000/40 M = 300 us < 500 us
```
Comment (yawhuu) [inline source/nielvisiii/academicIO.py]: The FPGA needs time to complete the configuration after we modify the configuration. 500us is the general time it takes to complete.

File: source/nielvisiii/academicIO.py [hunk 6]
```diff
@@ -425,15 +435,55 @@ def __init__(self, *configuration):
...
+        self.lsb_weights = 4882813.0 / 1E+9
+        self.offset = 0.0 / 1E+9
+        self.signed = True
+
+    def write(self, *args):
+        """
+        Writes values to one or more analog output channels. Use the
+        write(value) function to write a single point of data to the channel. Use
+        the write(values, sample_rate) to write multiple points of data to the
+        channel. The function does not return until the write to physical I/O
```
Comment (yawhuu) [inline source/nielvisiii/academicIO.py]: function means everything inside `def write():`. The program exists this function after the write completes.

File: examples/analog/AO_multipleChannels.py (hunk 4 above)
Comment (yawhuu) [inline examples/analog/AO_multipleChannels.py]:
```suggestion
        # write 1.0 and 2.5 to both AO0 and AO1 on bank A
```

File: source/nielvisiii/academicIO.py (hunk 5 above)
Comment (yawhuu) [inline source/nielvisiii/academicIO.py]:
```suggestion
            # after the configuration is modified, wait 500 us before applying the AI registers . 12*1000/40 M = 300 us < 500 us
```

File: source/nielvisiii/academicIO.py (hunk 6 above)
Comment (yawhuu) [inline source/nielvisiii/academicIO.py]:
```suggestion
        channel. The function hangs until the write to physical I/O
```

File: nielvis/academicIO.py [hunk 7]
```diff
@@ -444,19 +494,122 @@ def write(self, value):
...
+            max_write_samples = 10000
+
+            AnalogOutput.dma[bank].write(data[:max_write_samples], timeout_ms=0)
+            empty_elements_remaining = AnalogOutput.dma[bank].write([], timeout_ms=0)
+
+        def __wait_until_write_done(bank):
+            """
+            Configure IRQ and execute and wait until it is asserted. The
+            function will wait unitl the irq asserted which means the write to
+            physical I/O operation completes. The irq number is 31 Bank A and
```
Comment (ling-wang) [inline nielvis/academicIO.py]:
```suggestion
            physical I/O operation completes. The irq number is 31 for Bank A and
```

File: nielvis/academicIO.py [hunk 8]
```diff
@@ -444,19 +494,122 @@ def write(self, value):
...
+            
+            data_length = len(data)
+            max_write_samples = 10000
+
+            AnalogOutput.dma[bank].write(data[:max_write_samples], timeout_ms=0)
+            empty_elements_remaining = AnalogOutput.dma[bank].write([], timeout_ms=0)
+
+        def __wait_until_write_done(bank):
+            """
+            Configure IRQ and execute and wait until it is asserted. The
```
Comment (ling-wang) [inline nielvis/academicIO.py]:
```suggestion
            Configure IRQ, execute it, and wait until it is asserted,
```

File: nielvis/academicIO.py [hunk 9]
```diff
@@ -444,19 +494,122 @@ def write(self, value):
...
+            function will wait unitl the irq asserted which means the write to
```
Comment (ling-wang) [inline nielvis/academicIO.py]:
```suggestion
            which means the write to
```

File: nielvis/academicIO.py [hunk 10]
```diff
@@ -237,6 +238,9 @@ def __write_to_cnfg_register(bank, cnfg):
             while not(cnfg == self.cnfg[bank].read() and self.ready[bank].read()):
                 pass
 
+            # after the configuration is modified, wait 500 us before applying the AI registers . 12*1000/40 M = 300 us < 500 us
```
Comment (ling-wang) [inline nielvis/academicIO.py]:
```suggestion
            # after the configuration is modified, wait 500 us before applying the AI registers. 12*1000/40 M = 300 us < 500 us
```

File: PIPDESCRIPTION.md [hunk 11]
```diff
@@ -7,4 +7,7 @@
         <td>Author</td>
         <td>National Instruments</td>
     </tr>
-</table>
\ No newline at end of file
+</table>
+
+#### What's new:
+ - Analog Output N samples
```
Comment (ling-wang) [inline PIPDESCRIPTION.md]:
```suggestion
 - Analog Output N Samples
```

**Pull request merged**